### ISSUE TITLE
Eliminating node_modules whack-a-mole

### ISSUE THREAD

**Issue by pwaller** [alphagov/pp-development#50]
It turns out that the origin of a significant number of my problems is this:

```
$ bowl performance
I, [2014-03-06T11:08:44.338196 #3265]  INFO -- : Starting spotlight, data_out, backdrop_read, backdrop_write, backdrop_admin, backdrop, stagecraft, performance..
11:08:44 spotlight.1               | started with pid 3268
11:08:44 backdrop_read.1           | started with pid 3270
11:08:44 backdrop_write.1          | started with pid 3272
11:08:44 backdrop_admin.1          | started with pid 3276
11:08:44 stagecraft.1              | started with pid 3279
---- stuff happens ----
11:10:52 stagecraft.1              | exited with code 1
11:10:52 system                    | sending SIGTERM to all processes
11:10:52 spotlight.1               | terminated by SIGTERM
11:10:52 backdrop_read.1           | terminated by SIGTERM
11:10:52 backdrop_write.1          | terminated by SIGTERM
11:10:52 backdrop_admin.1          | terminated by SIGTERM
```

If one thing terminates, foreman sends a SIGTERM to everything else. Inside the `start-app.sh` for each of the applications they start by installing npm/python etc modules. When they get SIGTERM'd that leaves them in a corrupt state.

We believe this is the origin of the whack-a-mole game that everyone is familiar with.

The fix is to ensure that installation runs to completion, regardless if other applications fail.

@robyoung @alexmuller and @paulfurley all think that this might mean not using foreman (or at least not using it in this way).

**Comment by pwaller** [comment]
One possibility is to find a way of making `start-app.sh` safe under SIGTERM, in the sense that it should be prepared to receive one at any time (and not corrupt virtualenvs or node_modules). This may be tricky.

**Comment by robyoung** [comment]
This may be a good argument for separating dependency management from app running. Although that would raise the problem of how do we ensure dependencies are up to date when apps are run.

**Comment by pwaller** [comment]
Ah! This is actually likely not the fault of foreman exactly, since if you kill the bash script you don't kill the child (pip or npm), unless you exec. We know this because of #48. And (L)unix process behaviour.

The problem likely came from the standard operating procedure when you're in a borked situation (e.g, "port already in use") which is to kill the python processes or your VM. If foreman tore down but somewhere in the background NPM and friends are still going and then you send kill them, this might leave things in a bad state.

I suggest we introduce a cleanup script which deletes all of the module installation directories and document that if you're in a broken situation you should first try running that. Decoupling the installation is probably also wise. Thirdly, we should fix #48 so that when foreman exits the server processes are correctly killed.

**Comment by pwaller** [comment]
I think this issue can be closed in favour of #48 since that is the root cause.
